storage/database.go

```diff
@@ -158,34 +172,220 @@ func NewDatabase(
+
+		// if a namespace doesn't exist in newNamespaces, mark for removal
+		if err != nil {
+			removes = append(removes, ns.ID())
+			continue
+		}
+
+		// if namespace exists in newNamespaces, check if options are the same
+		optionsSame := newMd.Options().Equal(ns.Options())
+
+		// if options are the same, we don't need to do anything
+		if optionsSame {
+			continue
+		}
+
+		// if options are not the same, we mark for updates
+		updates = append(updates, newMd)
+	}
+
+	// check for any namespaces that need to be added
+	for _, ns := range newNamespaces.Metadatas() {
+		_, exists := d.namespaces[ns.ID().Hash()]
+		if !exists {
+			adds = append(adds, ns)
+		}
+	}
+
+	return removes, adds, updates
+}
+
+func (d *db) logNamespaceUpdate(removes []ts.ID, adds, updates []namespace.Metadata) error {
+	removalString, err := tsIDs(removes).String()
+	if err != nil {
+		return fmt.Errorf("unable to format removal, err = %v", err)
+	}
+
+	addString, err := metadatas(adds).String()
+	if err != nil {
+		return fmt.Errorf("unable to format adds, err = %v", err)
+	}
+
+	updateString, err := metadatas(updates).String()
+	if err != nil {
+		return fmt.Errorf("unable to format updates, err = %v", err)
+	}
+
+	// log scheduled operation
+	d.log.WithFields(
+		xlog.NewLogField("removals", removalString),
+		xlog.NewLogField("adds", addString),
+		xlog.NewLogField("updates", updateString),
+	).Infof("updating database namespaces")
+
+	return nil
+}
+
+func (d *db) addNamespacesWithLock(namespaces []namespace.Metadata) error {
 	for _, n := range namespaces {
-		if _, exists := ns[n.ID().Hash()]; exists {
-			return nil, errDuplicateNamespaces
+		// ensure namespace doesn't exist
+		_, ok := d.namespaces[n.ID().Hash()]
+		if ok { // should never happen
+			return fmt.Errorf("existing namespace marked for addition: %v", n.ID().String())
 		}
-		var blockRetriever block.DatabaseBlockRetriever
-		if blockRetrieverMgr != nil {
-			var newRetrieverErr error
-			blockRetriever, newRetrieverErr =
-				blockRetrieverMgr.Retriever(n.ID())
-			if newRetrieverErr != nil {
-				return nil, newRetrieverErr
-			}
+
+		// create and add to the database
+		newNs, err := d.newDatabaseNamespace(n)
+		if err != nil {
+			return err
 		}
-		ns[n.ID().Hash()] = newDatabaseNamespace(n, shardSet, blockRetriever,
+		d.namespaces[n.ID().Hash()] = newNs
+	}
+	return nil
+}
+
+func (d *db) newDatabaseNamespace(
+	md namespace.Metadata,
+) (databaseNamespace, error) {
+	brm := d.opts.DatabaseBlockRetrieverManager()
+	if brm == nil {
+		return newDatabaseNamespace(md, d.shardSet, nil,
 			d, d.writeCommitLogFn, d.opts)
 	}
-	d.namespaces = ns
 
-	mediator, err := newMediator(d,
-		opts.SetInstrumentOptions(iopts.SetMetricsScope(scope)))
+	blockRetriever, err := brm.Retriever(md)
 	if err != nil {
 		return nil, err
 	}
-	d.mediator = mediator
 
-	return d, nil
+	return newDatabaseNamespace(md, d.shardSet, blockRetriever,
+		d, d.writeCommitLogFn, d.opts)
+}
+
+type tsIDs []ts.ID
```

Comment:
Can you put `tsIDs` and `metadatas` type declaration and methods at the bottom of file?  Certain style guides grouping methods and then sorting by importance from top to bottom in a file.